DynamicData/Kernel/TwoStageRWLock.cs

```diff
@@ -0,0 +1,149 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+
+namespace DynamicData.Kernel
+{
+	/// <summary>
+	/// Similar to ReaderWriterLockSlim, but still allows reading after disposal.
+	/// All operations involving entering and exiting the lock in read mode still work after disposal.
+	/// All other operations will throw an NullReferenceException.
+	/// </summary>
+	/// <remarks>
+	/// Before disposing this object, all locks should be released.
+	/// Failure to do so will result in a SynchronizationLockException.
+	/// </remarks>
+	public class TwoStageRWLock : IDisposable
```

Comment:
This should be made internal otherwise I may end up having to support it to the public at large